File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/keyfederation/upload/DiagnosisKeysUploadService.java
package uk.nhs.nhsx.keyfederation.upload;

import com.amazonaws.services.lambda.runtime.Context;
import uk.nhs.nhsx.core.events.Events;
import uk.nhs.nhsx.core.events.InfoEvent;
import uk.nhs.nhsx.diagnosiskeydist.Submission;
import uk.nhs.nhsx.diagnosiskeydist.SubmissionRepository;
import uk.nhs.nhsx.keyfederation.BatchTagService;
import uk.nhs.nhsx.keyfederation.InteropClient;
import uk.nhs.nhsx.keyfederation.UploadedDiagnosisKeys;

import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import static uk.nhs.nhsx.core.Jackson.toJson;

public class DiagnosisKeysUploadService {
    private final static int NO_BATCH_SIZE_LIMIT_LEGACY_BATCH_SIZE = 0;

    private final InteropClient interopClient;
    private final SubmissionRepository submissionRepository;
    private final String region;
    private final BatchTagService batchTagService;
    private final boolean uploadRiskLevelDefaultEnabled;
    private final int uploadRiskLevelDefault;
    private final int initialUploadHistoryDays;
    private final int maxUploadBatchSize;
    private final int maxUploadBatchLimit;
    private final int maxSubsequentBatchUploadCount;
    private final Context context;
    private final Events events;
    private final Supplier<Instant> clock;

    public DiagnosisKeysUploadService(Supplier<Instant> clock,
                                      InteropClient interopClient,
                                      SubmissionRepository submissionRepository,
                                      BatchTagService batchTagService,
                                      String region,
                                      boolean uploadRiskLevelDefaultEnabled,
                                      int uploadRiskLevelDefault,
                                      int initialUploadHistoryDays,
                                      int maxUploadBatchSize,
                                      int maxSubsequentBatchUploadCount,
                                      Context context,
                                      Events events) {
        this.clock = clock;
        this.interopClient = interopClient;
        this.submissionRepository = submissionRepository;
        this.batchTagService = batchTagService;
        this.region = region;
        this.uploadRiskLevelDefaultEnabled = uploadRiskLevelDefaultEnabled;
        this.uploadRiskLevelDefault = uploadRiskLevelDefault;
        this.initialUploadHistoryDays = initialUploadHistoryDays;
        this.maxUploadBatchSize = maxUploadBatchSize;
        this.maxUploadBatchLimit = maxUploadBatchSize - 4; //4 mobile submissions/sec
        this.maxSubsequentBatchUploadCount = maxSubsequentBatchUploadCount;
        this.context = context;
        this.events = events;
    }

    public int loadKeysAndUploadToFederatedServer() throws Exception {
        int submissionCount = 0;
        long iterationDuration = 0L;
        Instant lastUploadedSubmissionTime = getLastUploadedTime();

        for (int i = 1; i <= maxSubsequentBatchUploadCount; i++) {
            var startTime = System.currentTimeMillis();
            var result = loadKeysAndUploadOneBatchToFederatedServer(lastUploadedSubmissionTime, i);

            submissionCount += result.submissionCount;

            if (maxUploadBatchSize == NO_BATCH_SIZE_LIMIT_LEGACY_BATCH_SIZE
                || lastUploadedSubmissionTime.equals(result.lastUploadedSubmissionTime)
                || result.submissionCount < maxUploadBatchLimit) {
                break;
            }

            lastUploadedSubmissionTime = result.lastUploadedSubmissionTime;
            iterationDuration = Math.max(iterationDuration, System.currentTimeMillis() - startTime);
            var remainingTimeInMillis = context.getRemainingTimeInMillis();
            if (iterationDuration >= remainingTimeInMillis) {
                break;
            }
        }

        return submissionCount;
    }

    static class BatchUploadResult {
        public final Instant lastUploadedSubmissionTime;
        public final int submissionCount;

        public BatchUploadResult(Instant lastUploadedSubmissionTime, int submissionCount) {
            this.lastUploadedSubmissionTime = lastUploadedSubmissionTime;
            this.submissionCount = submissionCount;
        }
    }

    public BatchUploadResult loadKeysAndUploadOneBatchToFederatedServer(Instant lastUploadedSubmissionTime, int batchNumber) throws Exception {

        events.emit(getClass(), new InfoEvent("Begin: Upload diagnosis keys to the Nearform server - batch " + batchNumber));

        List<Submission> newSubmissions = submissionRepository.loadAllSubmissions(
            lastUploadedSubmissionTime.toEpochMilli(),
            maxUploadBatchSize == NO_BATCH_SIZE_LIMIT_LEGACY_BATCH_SIZE ? Integer.MAX_VALUE : maxUploadBatchLimit,
            maxUploadBatchSize == NO_BATCH_SIZE_LIMIT_LEGACY_BATCH_SIZE ? Integer.MAX_VALUE : maxUploadBatchSize);

        List<ExposureUpload> exposureKeys = getUploadRequestRawPayload(newSubmissions);

        var transformedExposureKeys = exposureKeys.stream().map(this::updateRiskLevelIfDefaultEnabled).collect(Collectors.toList());

        events.emit(getClass(), new InfoEvent("Loading and transforming keys from submissions finished (from " +
            (newSubmissions.isEmpty() ? null : (newSubmissions.get(0).submissionDate)) +
            " to " + (newSubmissions.isEmpty() ? null : (newSubmissions.get(newSubmissions.size() - 1).submissionDate) )+
            "), keyCount=" + transformedExposureKeys.size() +
            " (batch " + batchNumber + ")"));

        if (!transformedExposureKeys.isEmpty()) {
            String payload = toJson(transformedExposureKeys);
            DiagnosisKeysUploadResponse uploadResponse = interopClient.uploadKeys(payload);
            if (uploadResponse != null) {
                events.emit(getClass(), new UploadedDiagnosisKeys(uploadResponse.insertedExposures, lastUploadedSubmissionTime, batchNumber));

                Instant updatedLastUploadedSubmissionTime = Instant.ofEpochMilli(newSubmissions.stream().map(submission -> submission.submissionDate.getTime()).max(Long::compare).orElseThrow());
                batchTagService.updateLastUploadState(updatedLastUploadedSubmissionTime.toEpochMilli() / 1000);

                return new BatchUploadResult(updatedLastUploadedSubmissionTime, newSubmissions.size());
            }
        } else {
            events.emit(getClass(), new InfoEvent("No keys were available for uploading to federation server with submission date greater than " +
                lastUploadedSubmissionTime +
                " -batch " + batchNumber));
        }

        return new BatchUploadResult(lastUploadedSubmissionTime, newSubmissions.size());
    }

    public ExposureUpload updateRiskLevelIfDefaultEnabled(ExposureUpload upload) {
        return new ExposureUpload(upload.keyData,
            upload.rollingStartNumber,
            uploadRiskLevelDefaultEnabled ? uploadRiskLevelDefault : upload.transmissionRiskLevel,
            upload.rollingPeriod,
            upload.regions);
    }

    public List<ExposureUpload> getUploadRequestRawPayload(List<Submission> submissions) // FIXME filter expired keys (rollingStartNumber & rollingPeriod)
    {
        List<ExposureUpload> exposures = new ArrayList<>();
        submissions
            .forEach(submission ->
                submission.payload.temporaryExposureKeys.forEach(exposure ->
                    exposures.add(
                        new ExposureUpload(
                            exposure.key,
                            exposure.rollingStartNumber,
                            exposure.transmissionRisk,
                            exposure.rollingPeriod,
                            List.of(this.region)
                        )
                    )));
        return exposures;
    }

    private Instant getLastUploadedTime() {
        return batchTagService.getLastUploadState()
            .map(it -> {
                events.emit(getClass(), new InfoEvent("Last uploaded timestamp from db " + it.lastUploadedTimeStamp));
                return Instant.ofEpochSecond(it.lastUploadedTimeStamp);
            })
            .orElse(
                clock.get().minus(Duration.ofDays(initialUploadHistoryDays))
            );
    }
}

File: src/aws/lambdas/incremental_distribution/cta/src/main/java/uk/nhs/nhsx/keyfederation/upload/DiagnosisKeysUploadService.kt
package uk.nhs.nhsx.keyfederation.upload

import com.amazonaws.services.lambda.runtime.Context
import uk.nhs.nhsx.core.Clock
import uk.nhs.nhsx.core.events.Events
import uk.nhs.nhsx.core.events.InfoEvent
import uk.nhs.nhsx.diagnosiskeydist.Submission
import uk.nhs.nhsx.diagnosiskeydist.SubmissionRepository
import uk.nhs.nhsx.keyfederation.BatchTagService
import uk.nhs.nhsx.keyfederation.DiagnosisKeysUploadIncomplete
import uk.nhs.nhsx.keyfederation.InteropClient
import uk.nhs.nhsx.keyfederation.InvalidTemporaryExposureKey
import uk.nhs.nhsx.keyfederation.UploadedDiagnosisKeys
import java.time.Duration
import java.time.Instant
import java.util.*
import kotlin.math.max

class DiagnosisKeysUploadService(
    private val clock: Clock,
    private val interopClient: InteropClient,
    private val submissionRepository: SubmissionRepository,
    private val batchTagService: BatchTagService,
    private val exposureUploadFactory: ExposureUploadFactory,
    private val uploadRiskLevelDefaultEnabled: Boolean,
    private val uploadRiskLevelDefault: Int,
    private val initialUploadHistoryDays: Int,
    private val maxUploadBatchSize: Int,
    private val maxSubsequentBatchUploadCount: Int,
    private val context: Context,
    private val events: Events
) {
    private val maxUploadBatchLimit: Int = maxUploadBatchSize - 4 // mobile submissions/sec

    fun loadKeysAndUploadToFederatedServer(): Int {
        var submissionCount = 0
        var iterationDuration = 0L
        var lastUploadedSubmissionTime = getLastUploadedTime()

        for (i in 1..maxSubsequentBatchUploadCount) {
            val startTime = clock().toEpochMilli()
            val result = loadKeysAndUploadOneBatchToFederatedServer(lastUploadedSubmissionTime, i)

            submissionCount += result.submissionCount

            if (maxUploadBatchSize == NO_BATCH_SIZE_LIMIT_LEGACY_BATCH_SIZE
                || lastUploadedSubmissionTime == result.lastUploadedSubmissionTime
                || result.submissionCount < maxUploadBatchLimit
            ) break

            lastUploadedSubmissionTime = result.lastUploadedSubmissionTime
            iterationDuration = max(iterationDuration, clock().toEpochMilli() - startTime)
            val remainingTimeInMillis = context.remainingTimeInMillis
            if (iterationDuration >= remainingTimeInMillis) break
        }

        return submissionCount
    }

    private fun loadKeysAndUploadOneBatchToFederatedServer(
        lastUploadedSubmissionTime: Instant,
        batchNumber: Int
    ): BatchUploadResult {
        events(InfoEvent("Begin: Upload diagnosis keys to the Nearform server - batch $batchNumber"))

        val newSubmissions = submissionRepository.loadAllSubmissions(
            lastUploadedSubmissionTime.toEpochMilli(),
            if (maxUploadBatchSize == NO_BATCH_SIZE_LIMIT_LEGACY_BATCH_SIZE) Int.MAX_VALUE else maxUploadBatchLimit,
            if (maxUploadBatchSize == NO_BATCH_SIZE_LIMIT_LEGACY_BATCH_SIZE) Int.MAX_VALUE else maxUploadBatchSize
        )

        val exposures = getUploadRequestRawPayload(newSubmissions)
            .filter { isKeyValid(it.keyData) }
            .map { updateRiskLevelIfDefaultEnabled(it) }

        events(
            InfoEvent("Loading and transforming keys from submissions finished (from ${if (newSubmissions.isEmpty()) null else newSubmissions[0].submissionDate} to ${if (newSubmissions.isEmpty()) null else newSubmissions[newSubmissions.size - 1].submissionDate}), keyCount=${exposures.size} (batch $batchNumber)")
        )

        if (exposures.isNotEmpty()) {
            val uploadResponse = interopClient.uploadKeys(exposures)

            emitStatistics(exposures, lastUploadedSubmissionTime, batchNumber)

            if (uploadResponse.insertedExposures != exposures.size) {
                events(
                    DiagnosisKeysUploadIncomplete(
                        exposures.size,
                        uploadResponse.insertedExposures,
                        lastUploadedSubmissionTime,
                        batchNumber
                    )
                )
            }

            val updatedLastUploadedSubmissionTime = newSubmissions
                .map { it.submissionDate.toEpochMilli() }
                .maxOrNull()
                ?.let(Instant::ofEpochMilli)
                ?: throw RuntimeException()

            batchTagService.updateLastUploadState(updatedLastUploadedSubmissionTime)

            return BatchUploadResult(
                updatedLastUploadedSubmissionTime,
                newSubmissions.size
            )
        } else {
            events(
                InfoEvent("No keys were available for uploading to federation server with submission date greater than $lastUploadedSubmissionTime -batch $batchNumber")
            )
        }

        return BatchUploadResult(
            lastUploadedSubmissionTime,
            newSubmissions.size
        )
    }

    private fun emitStatistics(
        exposures: List<ExposureUpload>,
        lastUploadedSubmissionTime: Instant,
        batchNumber: Int
    ) {
        exposures
            .groupBy(ExposureUpload::testType)
            .forEach { (testType, exposures) ->
                events(
                    UploadedDiagnosisKeys(
                        testType,
                        exposures.size,
                        lastUploadedSubmissionTime,
                        batchNumber
                    )
                )
            }
    }

    fun updateRiskLevelIfDefaultEnabled(upload: ExposureUpload): ExposureUpload = ExposureUpload(
        upload.keyData,
        upload.rollingStartNumber,
        if (uploadRiskLevelDefaultEnabled) uploadRiskLevelDefault else upload.transmissionRiskLevel,
        upload.rollingPeriod,
        upload.regions,
        upload.testType,
        upload.reportType,
        upload.daysSinceOnset
    )

    private fun isKeyValid(key: String?): Boolean {
        val isValid = key != null && isBase64EncodedAndEqualTo16Bytes(key)
        if (!isValid) events(InvalidTemporaryExposureKey(key))
        return isValid
    }

    private fun isBase64EncodedAndEqualTo16Bytes(value: String): Boolean = try {
        val decodedKeyData = Base64.getDecoder().decode(value)
        decodedKeyData.size == 16
    } catch (e: Exception) {
        false
    }

    // FIXME filter expired keys (rollingStartNumber & rollingPeriod)
    private fun getUploadRequestRawPayload(submissions: List<Submission>): List<ExposureUpload> = submissions
        .flatMap { exposureUploadFactory.create(it) }

    private fun getLastUploadedTime() = batchTagService
        .lastUploadState()
        ?.let {
            events(InfoEvent("Last uploaded timestamp from db ${it.lastUploadedTimeStamp}"))
            Instant.ofEpochSecond(it.lastUploadedTimeStamp)
        }
        ?: clock().minus(Duration.ofDays(initialUploadHistoryDays.toLong()))

    companion object {
        private const val NO_BATCH_SIZE_LIMIT_LEGACY_BATCH_SIZE = 0
    }
}
